src/charm.py

```diff
@@ -304,6 +306,27 @@ def _update_ingress_relation(self) -> None:
                 self._ingress_config, static=self._static_ingress_config
             )
 
+    def _update_tempo_api_relations(self) -> None:
+        """Update all applications related to us via the tempo-api relation."""
+        if not self.unit.is_leader():
+            return
+
+        # tempo-api should only send an external URL if it's set, otherwise it leaves that empty
+        internal_url = self._internal_url
+        external_url = self._external_url
+        if external_url == internal_url:
+            # external_url is not set and just defaulted back to internal_url.  Set it to None
+            external_url = None
+
+        tempo_api = TempoApiProvider(
```

Comment:
IMO we should strive to keep all relation wrappers inside `__init__`, that's the current design (for better or worse). The fact that we don't need to pass the charm to its constructor doesn't matter: we could do that here as well.

Reply:
I can make this change, but first I'll gently push back :D  

Agreed its a typical pattern to put relation handlers in init, but isn't that because they needed to `framework.observe`?  Putting these in `init` is the cause of this [bend-over-backwards-to-lazily-declare-things](https://github.com/canonical/tempo-coordinator-k8s-operator/pull/119#discussion_r1969334587) situation.  By instantiating `TempoApiProvider` here we eliminate the need to lazily define things

Reply:
After discussing offline, we are changing the object such that the constructor only takes invariate configurations (eg: relations, relation_name, and app) and then `publish` always requires the data to be published.  Eg:

```python
tempo_api = TempoApiProvider(
    relations=self.model.relations,
    relation_name=TEMPO_API_RELATION_ENDPOINT_NAME,
    app=self.app,
)

tempo_api.publish(urls...)
```

Then we'll put this constructor in the charm's init, but there's no worry about lazy declaration of things

Reply:
Done